### ISSUE TITLE
Stablize the dune describe command

### ISSUE THREAD

**Issue by hackwaly** [ocaml/dune#4263]
I'm integrating dune with ocamlearlybird. To make ocamlearlybird can discover debuggable executables. And get build informations as debug information in bytecode is not enough. I found dune describe is what I expected. However I got 
```
dune: Only --lang 0.1 is available at the moment as this command is not yet
      stabilised. If you would like to release a software that relies on the output
      of 'dune describe', please open a ticket on
      https://github.com/ocaml/dune.
```

**Comment by jeremiedimino** [comment]
Happy to help other tools in the community :) I've seen the ocamlearlybird post and that seems like an exciting project.

Regarding stabilising `dune describe`, we just introduced a versioned RPC in Dune and I'm wondering if we shouldn't focus on this effort rather than do two similar things in parallel. @rgrinberg what do you think of introducing requests to cover the things `dune describe` currently does?

@hackwaly is the fact that you can call `dune describe` as a one-off command important for ocamlearlybird?

**Comment by hackwaly** [comment]
I agree RPC is better method. And we prefer to move dune integration to vscode-ocaml-platform. So just add RPC method for discovering debuggable executable seems well.

**Comment by nojb** [comment]
For info, we are also using `dune describe` for some internal tools at LexiFi. Also, @nchataing recently developed a refactoring tool as part of an internship at LexiFi that also uses it https://github.com/nchataing/caml-migrate-floatarray.

Regarding using the RPC for this purpose, I have a few basic questions (they are more about the RPC mode in general, but am thinking about it from the point of view of our existing tools that use `dune describe`):
- does using the RPC imply using "watch mode" for compilation or the two can be used independently?
- Is the RPC protocol documented or is using the client library the only supported way of talking to the RPC server?

**Comment by jeremiedimino** [comment]
Ah, nice to know that this command is being useful to more projects :) It was initially a one hour train journey hack for ROTOR. And nice to know about `caml-migrate-floatarray`, that seems useful. I forwarded it to people who were looking at this inside Jane Street.

> does using the RPC imply using "watch mode" for compilation or the two can be used independently?

As the the tip of main, yes. But we can change that. For instance, there is a `dune rpc` command to establigh a RPC connection through stdin/stdout. We could add an option so that if the RPC is not running, we just performs the computations locally witthout going through the RPC. And if the one-off nature of `dune describe` is valuable, we could also add a command to do a single request. We could also keep `dune describe` as it and make it a thin wrapper around the RPC.

Basically, I'm thinking that if we start developing both `dune describe` and the RPC in parallel, then we are inevitably going to end up developing two parallel query languages and that doesn't feel right.

One consideration though; currently with `dune describe` you get everything at once. With the RPC, we might want to be more fine-grained. For instance, we might want one request to list the libraries, and further requests to get the details of each library. Same thing for executables. Or we could also do something a bit like graphql, where you issue a single request detailing everything you want. @emillon was looking at that at some point.

> Is the RPC protocol documented or is using the client library the only supported way of talking to the RPC server?

It is not documented at the moment, and the protocol might change until we release dune 3.0.0. But once it is stable, I see no reason not to document it.

**Comment by nojb** [comment]
> > does using the RPC imply using "watch mode" for compilation or the two can be used independently?
> 
> As the the tip of main, yes. But we can change that. For instance, there is a `dune rpc` command to establigh a RPC connection through stdin/stdout. We could add an option so that if the RPC is not running, we just performs the computations locally witthout going through the RPC. And if the one-off nature of `dune describe` is valuable, we could also add a command to do a single request. We could also keep `dune describe` as it and make it a thin wrapper around the RPC.

I think these all sound like reasonable features to have. Decoupling the RPC from the using watch mode for *compilation* may be a good idea. After all, using the RPC to query `dune`'s knowledge in a programmtic way seems pretty useful and independent from using it to drive an incremental build.

On the other hand, if the "fast" polling mode works as well as it sounds, then perhaps this will be a non-issue since we will be able to switch to it internally :)

> One consideration though; currently with `dune describe` you get everything at once. With the RPC, we might want to be more fine-grained. For instance, we might want one request to list the libraries, and further requests to get the details of each library. Same thing for executables. Or we could also do something a bit like graphql, where you issue a single request detailing everything you want. @emillon was looking at that at some point.

I think having a query language would be useful, but not sure how much work it would be to come up with a suitable spec :)

I guess whether more fine-grained requests make sense or not depends on the specific use made of this information. But a priori this seems like a good idea.

**Comment by jeremiedimino** [comment]
> On the other hand, if the "fast" polling mode works as well as it sounds, then perhaps this will be a non-issue since we will be able to switch to it internally :)

It of course will :)

> I think having a query language would be useful, but not sure how much work it would be to come up with a suitable spec :)

That's true. In any case, we can keep the current `dune describe` as it until we have a proper replacement. @nojb and @hackwaly I propose that you use (or continue using) `dune describe ` for now without the promise of long term of stability, and when we have a replacement we are happy to commit to maintaining long term in Dune, we can use this ticket to synchronise and ensure continuity of support for your use cases.

**Comment by rgrinberg** [comment]
> We could add an option so that if the RPC is not running, we just performs the computations locally witthout going through the RPC. And if the one-off nature of dune describe is valuable, we could also add a command to do a single request. We could also keep dune describe as it and make it a thin wrapper around the RPC.

Yeah, this is something that's on my TODO list. I don't quite know how to refactor the code so that such "local" connections are supported transparently without going through sockets or serializations, but it should be possible.

> Basically, I'm thinking that if we start developing both dune describe and the RPC in parallel, then we are inevitably going to end up developing two parallel query languages and that doesn't feel right.

It wouldn't be so bad. I would expect a `describe` rpc call to do exactly the same thing - except it will use csexp.

> It is not documented at the moment, and the protocol might change until we release dune 3.0.0. But once it is stable, I see no reason not to document it.

It is actually documented,  but there are no requests to replace what describe is currently doing. @nojb see `dune_rpc.mli` and `rpc.rst`.

> I think these all sound like reasonable features to have. Decoupling the RPC from the using watch mode for compilation may be a good idea. After all, using the RPC to query dune's knowledge in a programmtic way seems pretty useful and independent from using it to drive an incremental build.

My initial intention was to add an `--rpc` option to commands to allow the rpc server to be started and then shutdown after the command finishes, but we (with Jeremie) decided that it wasn't so well designed after all.  Another approach would be to add a flag to `init` like this: `dune rpc init --autostart`. This could try to connect to an existing instance if there's one running. If there's not, it could try and run it and terminate it after `dune rpc init` finishes.

**Comment by jeremiedimino** [comment]
> It wouldn't be so bad. I would expect a describe rpc call to do exactly the same thing - except it will use csexp.

To be more precise, I think it's fine to have multiple entry point for the same feature. What I'm more concerned about is multiplying the number of versioned languages we implement, because that's essentially widening our surface API and so increasing the maintenance load.
